server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java

```diff
@@ -37,9 +37,18 @@ public static DocumentMapper createEmpty(MapperService mapperService) {
         this.documentParser = documentParser;
         this.type = mapping.getRoot().name();
         this.mappingLookup = MappingLookup.fromMapping(mapping);
-        this.mappingSource = source;
-        assert mapping.toCompressedXContent().equals(source)
-            : "provided source [" + source + "] differs from mapping [" + mapping.toCompressedXContent() + "]";
+        if (sourceMapper().isSynthetic() && source.string().contains("\"_source\":{\"mode\":\"synthetic\"}") == false) {
```

Comment:
I think we can pass down an index version from the places a new `DocumentMapper` instance are created.
Maybe let's check the assertion only for tsdb indices before a version that have this fix?

Reply:
Done.